### ISSUE TITLE
[SLPVectorizer] Failure to keep complex math vectorized

### ISSUE THREAD

**Issue by RKSimon** [llvm/llvm-project#43353]
|  |  |
| --- | --- |
| Bugzilla Link | [44008](https://llvm.org/bz44008) |
| Version | trunk |
| OS | Windows NT |
| CC | @alexey-bataev,@anton-afanasyev,@topperc,@dtemirbulatov,@rotateright,@vporpo |

## Extended Description 
https://godbolt.org/z/YgBdFz
```c++
#include <complex>

std::complex<float> mul_f32(std::complex<float> &A) {
    return (2.0f * A);
}

std::complex<float> fma_f32(std::complex<float> &A, float B) {
    return (2.0f * A) + B;
}
```
for the simple mul case we manage to keep the real+imag components vectorized as a <2 x float>:

clang -g0 -O3 -march=btver2
```asm
mul_f32:
  vmovsd (%rdi), %xmm0 # xmm0 = mem[0],zero
  vaddps %xmm0, %xmm0, %xmm0
  retq
```
but with the fma case, we end up scalarizing:
```asm
fma_f32:
  vmovsd (%rdi), %xmm1 # xmm1 = mem[0],zero
  vmovshdup %xmm1, %xmm2 # xmm2 = xmm1[1,1,3,3]
  vaddss %xmm1, %xmm1, %xmm1
  vaddss %xmm2, %xmm2, %xmm2
  vaddss %xmm0, %xmm1, %xmm0
  vinsertps $16, %xmm2, %xmm0, %xmm0 # xmm0 = xmm0[0],xmm2[0],xmm0[2,3]
  retq
```
Ideally this would be:
```asm
fma_f32:
  vmovsd (%rdi), %xmm1 # xmm1 = mem[0],zero
  vaddps %xmm1, %xmm1, %xmm1
  vaddss %xmm0, %xmm1, %xmm1 # xmm1[0] += xmm0[0]
  vmovaps %xmm1, %xmm0
  retq
```

**Comment by vporpo** [comment]
This is not vectorized because fadd is missing from lane 0. This is the simplified input to SLP:
```llvm
define dso_local <2 x float> @fma_complex({{float, float}}* %A, float %B) {
entry:
  %0 = bitcast {{float, float}}* %A to <2 x float>*
  %1 = load <2 x float>, <2 x float>* %0
  %L0 = extractelement <2 x float> %1, i32 0
  %L1 = extractelement <2 x float> %1, i32 1
  %Mul0 = fmul float %L0, 2.000000e+00
  %Mul1 = fmul float %L1, 2.000000e+00
  %Fadd0 = fadd float %Mul0, %B
  %Ins1 = insertelement <2 x float> undef, float %Mul1, i32 1
  %Ins0 = insertelement <2 x float> %Ins1, float %Fadd0, i32 0
  ret <2 x float> %Ins0
}
```

**Comment by RKSimon** [comment]
Yes we don't have a common seed point any more, its annoying that we end up with this:

```llvm
  %L0 = extractelement <2 x float> %1, i32 0
  %L1 = extractelement <2 x float> %1, i32 1
  %Mul0 = fmul float %L0, 2.000000e+00
  %Mul1 = fmul float %L1, 2.000000e+00
```

and fail to vectorize it to:

```llvm
  %Mul = fmul <2 x float> %1, <2.000000e+00, 2.000000e+00>
  %L0 = extractelement <2 x float> %Mul, i32 0
  %L1 = extractelement <2 x float> %Mul, i32 1
```

I'm not sure what has load combined the <2 x float>? (It doesn't do it for the double case).

**Comment by vporpo** [comment]
I just noticed that there seems to be a bug in findBuildVector().

This generates a seed bundle: %Mul1, %Fadd0.

```llvm
  %Ins1 = insertelement <2 x float> undef, float %Mul1, i32 1
  %Ins0 = insertelement <2 x float> %Ins1, float %Fadd0, i32 0
```

But it should generate the reverse one: %Fadd0, %Mul1.

I think the elements simply follow the instruction order. They should be sorted by index. I will file a bug for this.

**Comment by alexey-bataev** [comment]
Partially fixed, the second example still must be addressed

**Comment by Andarwinux** [comment]
> Partially fixed, the second example still must be addressed

https://slp.godbolt.org/z/4e1vGMd7T

I noticed REVEC will break complex FMA vectorization with -fassociative-math -fno-signed-zeros -ffp-contract=fast.

**Comment by alexey-bataev** [comment]
> > Partially fixed, the second example still must be addressed
> 
> https://slp.godbolt.org/z/4e1vGMd7T
> 
> I noticed REVEC will break complex FMA vectorization with -fassociative-math -fno-signed-zeros -ffp-contract=fast.

Better to create a separate issue rather than adding it here

**Comment by Andarwinux** [comment]
> > > Partially fixed, the second example still must be addressed
> > 
> > 
> > https://slp.godbolt.org/z/4e1vGMd7T
> > I noticed REVEC will break complex FMA vectorization with -fassociative-math -fno-signed-zeros -ffp-contract=fast.
> 
> Better to create a separate issue rather than adding it here

https://github.com/llvm/llvm-project/issues/199715

**Comment by alexey-bataev** [comment]
This issue is considered to be fixed, but insertvalue sequences currently are not vectorized, since there is no quick conversion from vector to struct. Currently, perform this by vector store/struct load, which is expensive and does not allow to vectorize insertvalues effectively
